library/std/src/sys/xous/thread_local_key.rs

```diff
@@ -23,61 +23,66 @@ pub type Dtor = unsafe extern "C" fn(*mut u8);
 
 const TLS_MEMORY_SIZE: usize = 4096;
 
-/// TLS keys start at `1` to mimic POSIX.
+/// TLS keys start at `1`. Index `0` is unused
+#[cfg_attr(test, linkage = "available_externally")]
```

Comment:
This is LLVM specific. Could you instead use an extern block for the test case?

```rust
#[cfg(not(test))]
#[export_name = "_ZN16__rust_internals3std3sys4xous16thread_local_key13TLS_KEY_INDEXE"]
static TLS_KEY_INDEX: AtomicUsize = AtomicUsize::new(1);

#[cfg(test)]
extern {
    #[export_name = "_ZN16__rust_internals3std3sys4xous16thread_local_key13TLS_KEY_INDEXE"]
    static TLS_KEY_INDEX: AtomicUsize;
}
```

Reply:
Also https://github.com/rust-lang/rust/blob/ddca5343f2988f49fa407dc1ff43bcf779ea4173/library/std/src/thread/mod.rs#L196-L212 should cause the TLS implementation of realstd rather than the libstd being tested to be used. Are you trying to use this module somewhere in a way that isn't intercepted by this code?

Reply:
That's what I had originally, but I ended up copying it from SGX:

https://github.com/rust-lang/rust/blob/ddca5343f2988f49fa407dc1ff43bcf779ea4173/library/std/src/sys/sgx/abi/tls/mod.rs#L15-L24

I'll swap it back here and in the other places I use it.

The problem with using an extern block is that I need to mark it `unsafe {}` all over the place. Which is fine, it just adds more noise.

Reply:
I'm not sure what path is being taken, but I was noticing that keys were getting reused. Also, destructors weren't getting run. Forcing them to be the same solved the issue.

Reply:
I think https://github.com/rust-lang/rust/blob/ddca5343f2988f49fa407dc1ff43bcf779ea4173/library/std/src/sys/xous/thread.rs#L80 needs to call `realstd::sys::thread_local_key::destroy_tls()` when testing.

Reply:
> That's what I had originally, but I ended up copying it from SGX:

I noticed that they do this when I looked into how other targets were handling the system allocator.

> The problem with using an extern block is that I need to mark it unsafe {} all over the place. Which is fine, it just adds more noise.

Makes sense.

If my above comment fixes the TLS issue then great. If not keeping the code you have right now with externally_available is fine by me for now. It would be nice to get it to compile with GCC and Cranelift, but it is not all that important as Xous only runs on riscv anyway, which LLVM fully supports.

Reply:
I did have a version of the patch where I stored the destructor in the TLS table itself (at offset 0) and called that. But I still ran into a problem where threads spawn with the main libstd and do things like create locks for stdout, then the actual test runs with the test libstd and gets a brand new copy of `TLS_KEY_INDEX` so it proceeds to reuse keys.

There were two problems: `DTORS` wasn't getting shared, and `TLS_KEY_INDEX` was letting keys get reused.

The "extern" trick solves both problems.

Reply:
On other targets `crate::sys::thread_local_keys` is entirely unused when testing from what I can tell. Instead the entire tls implementation of "realstd" is used.

Reply:
I'm not sure what's causing it to not do that, but I do see two different versions of the symbols. Additionally, without this patch it gets stuck on `thread::local::tests::dtors_in_dtors_in_dtors ` because the destructor isn't running properly:

```
0028f10c d _ZN3std3sys4xous16thread_local_key13TLS_KEY_INDEX17h0ab63fd833369238E
0028f12c d _ZN3std3sys4xous16thread_local_key13TLS_KEY_INDEX17hf9c392ed3fb3ec68E
```

I'll change both this and `DLMALLOC` to be `extern "Rust"` instead of using the llvm trick.